rust-arroyo/src/backends/kafka/producer.rs

```diff
@@ -98,13 +109,21 @@ impl ArroyoProducer<KafkaPayload> for KafkaProducer {
         destination: &TopicOrPartition,
         payload: KafkaPayload,
     ) -> Result<(), ProducerError> {
-        let base_record = payload.to_base_record(destination);
+        let (tx, rx) = channel::<Result<(), KafkaError>>();
+        let base_record = payload.to_base_record(destination, Arc::new(tx));
 
+        // If the producer fails to enqueue the message, this will return an error
         self.producer
             .send(base_record)
             .map_err(|(kafka_error, _record)| ProducerError::from(kafka_error))?;
 
-        Ok(())
+        // If the producer fails to flush the message out of the buffer, the delivery callback will send an error
+        // on the channel.
+        match rx.recv_timeout(Duration::from_secs(5)) {
+            Ok(Ok(_)) => Ok(()),
+            Ok(Err(e)) => Err(ProducerError::from(e)),
+            Err(_) => Err(ProducerError::ProduceWaitTimeout),
+        }
```

Comment:
```suggestion
        rx.recv_timeout(Duration::from_secs(5))
             .map_err(|_| ProducerError::ProduceWaitTimeout)?
             .map_err(ProducerError::from)
```

(untested)

Reply:
but also not sure, do we want this 5 second timeout?

Reply:
I don't think we want it to block indefinitely. I'm open to suggestions though.